### PR TITLE
Word reordering errors analysis

### PR DESCRIPTION
**Pull request by zdou0830**
Compute Kendall's tau distances to analyze word reordering errors #16 .

Example usage:

python compare_mt.py example/ted.ref.eng example/ted.sys1.eng example/ted.sys2.eng  --src_file example/ted.orig.slk --compare_word_reordering_errors ref_align=example/ted.r    ef.align,out1_align=example/ted.sys1.align,out2_align=example/ted.sys2.align

### CODE REVIEW

**Comment by neubig** [comment]
Thanks! This seems like a good implementation of the paper you cited in the issue, but I'm actually not sure if their way of doing things is the best... I actually prefer the idea behind RIBES, which is that you find alignments directly between the reference and the outputs. The reason why I think this is better is because it should be significantly easier to find alignments between sentences in the same language than between sentences in different languages, so I think it will be much less prone to error. What do you think?

**Comment by zdou0830** [comment]
Thank you! I think both methods have their pros and cons. Finding alignments between sentences in the same language might be easier but we may have less training data. I can try to implement RIBES too.

**Comment by neubig** [comment]
Thanks! Just to be clear: RIBES doesn't even use training data to find alignments. It just calculates exact matches between words (with some heuristics to resolve ambiguity if a single word appears multiple times), and calculates the Kendall's tau among the matched words.

**Comment by zdou0830** [comment]
There are some slight differences between the original paper of RIBES (http://www.anthology.aclweb.org/D/D10/D10-1092.pdf) and their latest implementations. It doesn't make huge differences, and here I follow the original paper. (They switch line 2 and line 3 of the for-loop in Figure 1.)

Maybe we can delete the previous method and just keep the RIBES part.

File: ribes_utils.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import itertools
```
Comment (neubig) [inline ribes_utils.py]: I think RIBES should be a `Scorer`. Could you move it there and have it follow the same format as, for example, `BleuScorer`? This would allow it to be re-used in other parts of the code.

File: ribes_utils.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    gram_pos[i+1] = defaultdict(lambda: [])
+
+  for i, word in enumerate(sent):
+    for j in range(min(i+1, order)):
+      gram_pos[j+1][word].append(i-j)
+      word = sent[i-j-1] + ' ' + word
+  
+  return gram_pos
+
+def word_alignment(ref, out, order=2):
```
Comment (neubig) [inline ribes_utils.py]: I think this is good because it basically resolves https://github.com/neulab/compare-mt/issues/17

Maybe we could split this function off into a file `align_utils` and have it be `ngram_context_align` or something? Then you could have a comment indicating that this is the algorithm proposed by RIBES. Then `RibesScorer` would call this function to generate alignments, but we could also use it for other things, such as the word accuracy calculations (which are using a different implicit alignment algorithm).

File: compare_mt.py [hunk 3]
```diff
@@ -229,6 +231,38 @@ def print_ngram_report(ref, out1, out2,
     print('{}\t{} (sys1={}, sys2={})'.format(' '.join(k), v, match1[k], match2[k]))
   print()
 
+def print_word_reorder_error_report(ref, out1, out2,
```
Comment (neubig) [inline compare_mt.py]: I think that once we create `RibesScorer` (see comment below), we can remove this function altogether.

File: align_utils.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import itertools
```
Comment (neubig) [inline align_utils.py]: I'm not sure what to do with this, but maybe it can be deleted if it's not used anywhere?

**Comment by neubig** [comment]
Great, thanks!

**Comment by alphadl** [comment]
Is this "compare reordering errors" module still available in the master branch?

**Comment by zdou0830** [comment]
> Is this "compare reordering errors" module still available in the master branch?

Yes, you can compare reordering errors by using the Ribes scorer (--compare_scores 'score_type=ribes' ).
